Build OpenWRT on MacOS arm64 (Apple Silicon)

svlobanov opened the pull request:
This PR contains three patches that allow to build openwrt on MacOS arm64 (macOS 11.6, Xcode 13.0 build 13A233, kernel 20.6.0 arm64)

Build instructions updated on wiki page https://openwrt.org/docs/guide-developer/toolchain/buildroot.exigence.macosx

ptpt52 commented:
great!

ptpt52 commented:
build for ipq40xx error:
```
rm -f /Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/x-wrt-21.10-b202111291458-ipq40xx-generic-alfa-network_ap120c-ac-initramfs-fit-uImage.itb
cp /Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/Image-initramfs /Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/x-wrt-21.10-b202111291458-ipq40xx-generic-alfa-network_ap120c-ac-initramfs-fit-uImage.itb
gzip -f -9n -c /Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/x-wrt-21.10-b202111291458-ipq40xx-generic-alfa-network_ap120c-ac-initramfs-fit-uImage.itb  > /Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/x-wrt-21.10-b202111291458-ipq40xx-generic-alfa-network_ap120c-ac-initramfs-fit-uImage.itb.new
/Volumes/x-wrt/x-wrt/scripts/mkits.sh -D alfa-network_ap120c-ac -o /Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/x-wrt-21.10-b202111291458-ipq40xx-generic-alfa-network_ap120c-ac-initramfs-fit-uImage.itb.its -k /Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/x-wrt-21.10-b202111291458-ipq40xx-generic-alfa-network_ap120c-ac-initramfs-fit-uImage.itb -C gzip   -d /Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-5.10.82/arch/arm/boot/dts/qcom-ipq4018-ap120c-ac.dtb   -a 0x80208000 -e 0x80208000    -c config@1 -A arm -v 5.10.82
PATH=/Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-5.10.82/scripts/dtc:/Volumes/x-wrt/x-wrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-11.2.0_musl_eabi/bin:/Volumes/x-wrt/x-wrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-11.2.0_musl_eabi/bin:/Volumes/x-wrt/x-wrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-11.2.0_musl_eabi/bin:/Volumes/x-wrt/x-wrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-11.2.0_musl_eabi/bin:/Volumes/x-wrt/x-wrt/staging_dir/host/bin:/Volumes/x-wrt/x-wrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-11.2.0_musl_eabi/bin:/Volumes/x-wrt/x-wrt/staging_dir/host/bin:/Volumes/x-wrt/x-wrt/staging_dir/host/bin:/Users/ptpt52/vmroot/opt/findutils/libexec/gnubin:/Users/ptpt52/vmroot/Cellar/coreutils/9.0/libexec/gnubin:/Users/ptpt52/vmroot/opt/gettext/bin:/Users/ptpt52/vmroot/opt/gnu-getopt/bin:/Users/ptpt52/vmroot/opt/make/libexec/gnubin:/Users/ptpt52/vmroot/bin:/Users/ptpt52/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin mkimage  -f /Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/x-wrt-21.10-b202111291458-ipq40xx-generic-alfa-network_ap120c-ac-initramfs-fit-uImage.itb.its /Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/x-wrt-21.10-b202111291458-ipq40xx-generic-alfa-network_ap120c-ac-initramfs-fit-uImage.itb.new
make[5]: *** [Makefile:18: /Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/x-wrt-21.10-b202111291458-ipq40xx-generic-alfa-network_ap120c-ac-initramfs-fit-uImage.itb] Segmentation fault: 11
make[5]: *** Deleting file '/Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/x-wrt-21.10-b202111291458-ipq40xx-generic-alfa-network_ap120c-ac-initramfs-fit-uImage.itb'
make[5]: Leaving directory '/Volumes/x-wrt/x-wrt/target/linux/ipq40xx/image'
make[4]: *** [Makefile:24: install] Error 2
make[4]: Leaving directory '/Volumes/x-wrt/x-wrt/target/linux/ipq40xx'
make[3]: *** [Makefile:11: install] Error 2
make[3]: Leaving directory '/Volumes/x-wrt/x-wrt/target/linux'
time: target/linux/install#36.86#14.39#64.39
    ERROR: target/linux failed to build.
make[2]: *** [target/Makefile:30: target/linux/install] Error 1
make[2]: Leaving directory '/Volumes/x-wrt/x-wrt'
make[1]: *** [target/Makefile:24: /Volumes/x-wrt/x-wrt/staging_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/stamp/.target_install] Error 2
make[1]: Leaving directory '/Volumes/x-wrt/x-wrt'
make: *** [/Volumes/x-wrt/x-wrt/include/toplevel.mk:230: world] Error 2
```

ptpt52 commented:
mkimage Segmentation fault: 11
```
PATH=/Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-5.10.82/scripts/dtc:/Volumes/x-wrt/x-wrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-11.2.0_musl_eabi/bin:/Volumes/x-wrt/x-wrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-11.2.0_musl_eabi/bin:/Volumes/x-wrt/x-wrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-11.2.0_musl_eabi/bin:/Volumes/x-wrt/x-wrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-11.2.0_musl_eabi/bin:/Volumes/x-wrt/x-wrt/staging_dir/host/bin:/Volumes/x-wrt/x-wrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-11.2.0_musl_eabi/bin:/Volumes/x-wrt/x-wrt/staging_dir/host/bin:/Volumes/x-wrt/x-wrt/staging_dir/host/bin:/Users/ptpt52/vmroot/opt/findutils/libexec/gnubin:/Users/ptpt52/vmroot/Cellar/coreutils/9.0/libexec/gnubin:/Users/ptpt52/vmroot/opt/gettext/bin:/Users/ptpt52/vmroot/opt/gnu-getopt/bin:/Users/ptpt52/vmroot/opt/make/libexec/gnubin:/Users/ptpt52/vmroot/bin:/Users/ptpt52/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin mkimage  -f /Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/x-wrt-21.10-b202111291458-ipq40xx-generic-alfa-network_ap120c-ac-initramfs-fit-uImage.itb.its /Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/x-wrt-21.10-b202111291458-ipq40xx-generic-alfa-network_ap120c-ac-initramfs-fit-uImage.itb.new

```

aparcar commented:
Hi this is great! Also an awesome round trip time since I posted the issue just this afternoon!

Please add commit messages briefly describing why this is needed, it's a requirement for all commits to have a short message

svlobanov commented:
> mkimage Segmentation fault: 11
> 
> ```
> PATH=/Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/linux-5.10.82/scripts/dtc:/Volumes/x-wrt/x-wrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-11.2.0_musl_eabi/bin:/Volumes/x-wrt/x-wrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-11.2.0_musl_eabi/bin:/Volumes/x-wrt/x-wrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-11.2.0_musl_eabi/bin:/Volumes/x-wrt/x-wrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-11.2.0_musl_eabi/bin:/Volumes/x-wrt/x-wrt/staging_dir/host/bin:/Volumes/x-wrt/x-wrt/staging_dir/toolchain-arm_cortex-a7+neon-vfpv4_gcc-11.2.0_musl_eabi/bin:/Volumes/x-wrt/x-wrt/staging_dir/host/bin:/Volumes/x-wrt/x-wrt/staging_dir/host/bin:/Users/ptpt52/vmroot/opt/findutils/libexec/gnubin:/Users/ptpt52/vmroot/Cellar/coreutils/9.0/libexec/gnubin:/Users/ptpt52/vmroot/opt/gettext/bin:/Users/ptpt52/vmroot/opt/gnu-getopt/bin:/Users/ptpt52/vmroot/opt/make/libexec/gnubin:/Users/ptpt52/vmroot/bin:/Users/ptpt52/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin mkimage  -f /Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/x-wrt-21.10-b202111291458-ipq40xx-generic-alfa-network_ap120c-ac-initramfs-fit-uImage.itb.its /Volumes/x-wrt/x-wrt/build_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/linux-ipq40xx_generic/tmp/x-wrt-21.10-b202111291458-ipq40xx-generic-alfa-network_ap120c-ac-initramfs-fit-uImage.itb.new
> ```

Could you please provide your subtarget and target profile?

svlobanov commented:
> Hi this is great! Also an awesome round trip time since I posted the issue just this afternoon!
> 
> Please add commit messages briefly describing why this is needed, it's a requirement for all commits to have a short message

Thank you for your review. Commit messages have been added.

ptpt52 commented:
and fixup needed for `package/boot/uboot-fritz4040/`
```
diff --git a/tools/dumpimage.c b/tools/dumpimage.c
index 99c0d0e..56d918c 100644
--- a/tools/dumpimage.c
+++ b/tools/dumpimage.c
@@ -7,6 +7,7 @@
  */
 
 #include "dumpimage.h"
+#include "sysupgrade.h"
 #include <image.h>
 #include <version.h>
```

ptpt52 commented:
same target build on linux is ok.

aparcar commented:
I just checked and `mkimage` is indeed broken. The other two patches work fine. I'll merge the working code

svlobanov commented:
> I just checked and `mkimage` is indeed broken. The other two patches work fine. I'll merge the working code

Ok, I will work with upstream to fix mkimage on apple arm64. Do I need to remove commit related to mkimage from this PR?

aparcar commented:
> Ok, I will work with upstream to fix mkimage on apple arm64

Thanks!

> Do I need to remove commit related to mkimage from this PR?

No I can take care of that

svlobanov commented:
I've started a discussion related to mkimage issue in upstream (u-boot) mailing list. https://lists.denx.de/pipermail/u-boot/2021-November/468564.html

svlobanov commented:
@aparcar @ptpt52 I've found workaround for mkimage issue.

The root cause of mkimage segfault is ASLR mechanism enabled on modern MacOS systems. The only way to disable ASLR is run application from debugger (LLDB). I've made a very simple wrapper based on LLDB source code to disable ASLR. It is uploaded to github https://github.com/svlobanov/disable-aslr-on-mac

After wrapping mkimage like it is described in README file above, I've finally built OpenWRT for ipq40xx

If proposed solution is acceptable for OpenWRT project I can prepare a patch for OpenWRT to include and use this wrapper (only) on MacOS systems.

```
sergey@Sergeys-MacBook-Air generic % pwd         
/Volumes/OpenWrt/openwrt3/bin/targets/ipq40xx/generic
sergey@Sergeys-MacBook-Air generic % ls -la | grep ipq40xx
-rw-r--r-- 1 sergey staff 8294596 Dec  1 22:47 openwrt-ipq40xx-generic-alfa-network_ap120c-ac-initramfs-fit-uImage.itb
-rw-r--r-- 1 sergey staff 4230793 Dec  1 22:47 openwrt-ipq40xx-generic-alfa-network_ap120c-ac-rootfs.tar.gz
-rw-r--r-- 1 sergey staff 7996751 Dec  1 22:47 openwrt-ipq40xx-generic-alfa-network_ap120c-ac-squashfs-nand-factory.bin
-rw-r--r-- 1 sergey staff 7363376 Dec  1 22:47 openwrt-ipq40xx-generic-alfa-network_ap120c-ac-squashfs-nand-sysupgrade.bin
-rw-r--r-- 1 sergey staff    3344 Dec  1 22:47 openwrt-ipq40xx-generic-alfa-network_ap120c-ac.manifest
```

neheb commented:
Needing to disable ASLR sounds like a bug.

ptpt52 commented:
what is ASLR------^_^

svlobanov commented:
> what is ASLR------^_^

https://en.wikipedia.org/wiki/Address_space_layout_randomization @ptpt52  I hope you will be able to open this page if you are in China

svlobanov commented:
Link to homebrew issue https://github.com/Homebrew/homebrew-core/issues/90278

k-ronny commented:
@svlobanov I found a workaround just for me. See my comments at [https://github.com/u-boot/u-boot/commit/3b142045e8a7f0ab17b6099e9226296af45967d0#diff-c7afb0e15b9fd57ca821cf14449af4d9432f73b8fffc01432a23d9fc27fb8542](https://github.com/u-boot/u-boot/commit/3b142045e8a7f0ab17b6099e9226296af45967d0#diff-c7afb0e15b9fd57ca821cf14449af4d9432f73b8fffc01432a23d9fc27fb8542).

aparcar commented:
@nbd168 do you have an opinion on these changes?

svlobanov commented:
@aparcar As I see, @k-ronny found at least two solutions to make mkimage compatible with MacOS without aslr disabling dirty hack.

First solution uses "undocumented" function getsectiondata. It is mentioned in Apple release notes https://opensource.apple.com/source/cctools/cctools-795/libmacho/notes.auto.html. Also, it is documented in dlang documention http://www.dpldocs.info/experimental-docs/core.sys.darwin.mach.getsect.getsectiondata.html

I will check both solution and propose a patch to OpenWRT for mkimage, after that will try to work with upstream but as I understand, MacOS compatibility is not a priority for most u-boot developers.

nbd168 commented:
Does this help? https://source.denx.de/u-boot/u-boot/-/commit/3b142045e8a7f0ab17b6099e9226296af45967d0

svlobanov commented:
> u-boot/u-boot@3b14204#diff-c7afb0e15b9fd57ca821cf14449af4d9432f73b8fffc01432a23d9fc27fb8542

this patch allows compiling mkimage on macos, but mkimage segfaults, I have included this patch to the PR, but @ptpt52 
 observed that mkimage segfaults

@k-ronny suggested two another patches here https://github.com/u-boot/u-boot/commit/3b142045e8a7f0ab17b6099e9226296af45967d0#diff-c7afb0e15b9fd57ca821cf14449af4d9432f73b8fffc01432a23d9fc27fb8542

I will check both

ptpt52 commented:
another issue, may not be related here but build on macos m1 failed:
```
echo "libtirpc" >> /Volumes/x-wrt/x-wrt/staging_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/pkginfo/libtirpc.default.install
make[2]: Leaving directory '/Volumes/x-wrt/x-wrt/feeds/packages/libs/libtirpc'
time: package/feeds/packages/libtirpc/compile#0.10#0.11#0.25
make[2]: Entering directory '/Volumes/x-wrt/x-wrt/feeds/packages/libs/libtirpc'
make[2]: Leaving directory '/Volumes/x-wrt/x-wrt/feeds/packages/libs/libtirpc'
time: package/feeds/packages/libtirpc/host-compile#0.09#0.10#0.22
make[2]: Entering directory '/Volumes/x-wrt/x-wrt/feeds/packages/net/nfs-kernel-server'
(cd /Volumes/x-wrt/x-wrt/build_dir/hostpkg/nfs-utils-2.5.4/; if [ -x configure ]; then cp -fpR /Volumes/x-wrt/x-wrt/scripts/config.{guess,sub} /Volumes/x-wrt/x-wrt/build_dir/hostpkg/nfs-utils-2.5.4// && CC="gcc" CFLAGS="-O2 -I/Volumes/x-wrt/x-wrt/staging_dir/host/include -I/Volumes/x-wrt/x-wrt/staging_dir/hostpkg/include -I/Volumes/x-wrt/x-wrt/staging_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/host/include -Dlinux -Wno-error=missing-include-dirs" CXX="g++" CPPFLAGS="-I/Volumes/x-wrt/x-wrt/staging_dir/host/include -I/Volumes/x-wrt/x-wrt/staging_dir/hostpkg/include -I/Volumes/x-wrt/x-wrt/staging_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/host/include" LDFLAGS="-L/Volumes/x-wrt/x-wrt/staging_dir/host/lib -L/Volumes/x-wrt/x-wrt/staging_dir/hostpkg/lib -L/Volumes/x-wrt/x-wrt/staging_dir/target-arm_cortex-a7+neon-vfpv4_musl_eabi/host/lib" CONFIG_SHELL="/usr/bin/env bash" ac_cv_lib_event_event_dispatch=yes ac_cv_lib_nfsidmap_nfs4_init_name_mapping=yes ac_cv_lib_blkid_blkid_get_library_version=yes ac_cv_header_event_h=yes ac_cv_header_nfsidmap_h=yes ac_cv_header_blkid_blkid_h=yes ac_cv_lib_resolv___res_querydomain=yes GSSGLUE_CFLAGS=" " GSSGLUE_LIBS=" " RPCSECGSS_CFLAGS=" " RPCSECGSS_LIBS=" " CONFIG_SQLITE3_TRUE="#" CONFIG_NFSDCLD_TRUE="#"  bash ./configure --target=aarch64-apple-darwin21.1.0 --host=aarch64-apple-darwin21.1.0 --build=aarch64-apple-darwin21.1.0 --program-prefix="" --program-suffix="" --prefix=/Volumes/x-wrt/x-wrt/staging_dir/hostpkg --exec-prefix=/Volumes/x-wrt/x-wrt/staging_dir/hostpkg --sysconfdir=/Volumes/x-wrt/x-wrt/staging_dir/hostpkg/etc --localstatedir=/Volumes/x-wrt/x-wrt/staging_dir/hostpkg/var --sbindir=/Volumes/x-wrt/x-wrt/staging_dir/hostpkg/bin --disable-gss --disable-nfsv4 --disable-nfsv41 --without-tcp-wrappers --with-rpcgen=internal ; fi )
configure: loading site script /Volumes/x-wrt/x-wrt/include/site/darwin
checking build system type... aarch64-apple-darwin21.1.0
checking host system type... aarch64-apple-darwin21.1.0
checking for a BSD-compatible install... /Volumes/x-wrt/x-wrt/staging_dir/host/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /Users/ptpt52/vmroot/opt/coreutils/libexec/gnubin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking whether make supports the include directive... yes (GNU style)
checking for aarch64-apple-darwin21.1.0-gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... gcc3
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /Volumes/x-wrt/x-wrt/staging_dir/host/bin/grep
checking for egrep... /Volumes/x-wrt/x-wrt/staging_dir/host/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking minix/config.h usability... no
checking minix/config.h presence... no
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking pkg-config is at least version 0.9.0... yes
checking for TIRPC... yes
checking for authgss_free_private_data in -ltirpc... no
checking for libtirpc_set_debug in -ltirpc... yes
checking for prctl... no
configure: error: prctl syscall is not available
make[2]: *** [Makefile:207: /Volumes/x-wrt/x-wrt/build_dir/hostpkg/nfs-utils-2.5.4/.configured] Error 1
make[2]: Leaving directory '/Volumes/x-wrt/x-wrt/feeds/packages/net/nfs-kernel-server'
time: package/feeds/packages/nfs-kernel-server/host-compile#0.93#0.74#2.98
    ERROR: package/feeds/packages/nfs-kernel-server [host] failed to build.
make[1]: *** [package/Makefile:114: package/feeds/packages/nfs-kernel-server/host/compile] Error 1
make[1]: Leaving directory '/Volumes/x-wrt/x-wrt'
make: *** [/Volumes/x-wrt/x-wrt/include/toplevel.mk:230: package/feeds/packages/nfs-kernel-server/compile] Error 2
```

svlobanov commented:
> another issue, may not be related here but build on macos m1 failed: `make V=s package/feeds/packages/nfs-kernel-server/compile`

This issue is not related to this repository (https://github.com/openwrt/openwrt ). Please open an issue in https://github.com/openwrt/packages . That repo has its own PRs and issue tracker

aparcar commented:
@nbd168 if you don't object I merge the patches for gcc and fakeroot.

nbd168 commented:
No objection from me

svlobanov commented:
@aparcar  thank you for merging. I'm moving this PR to WIP status to fix mkimage issue

svlobanov commented:
> Please rebase this commit. You want it to only contain the third (unmerged) commit

Sorry for my mistake. I will report to you when it's done

svlobanov committed:
tools/mkimage: fix build on MacOS arm64

Fixed -no-pie compilation warning on MacOS
Fixed errors related to using absolute addressing on MacOS arm64

Based on upstream patch from Jessica Clarke and suggestions from Ronny Kotzschmar

Link to original patch and discussion:
https://github.com/u-boot/u-boot/commit/3b142045e8a7f0ab17b6099e9226296af45967d0

Signed-off-by: Sergey V. Lobanov <sergey@lobanov.in>

svlobanov commented:
@aparcar I have prepared a new patch and tested build. it works on ipq40xx target (mkimage doesn't segfault)

ptpt52 commented:
changes is good, but I prefer do some small change:
```
diff --git a/tools/imagetool.h b/tools/imagetool.h
index 6ae662a..da317e4 100644
--- a/tools/imagetool.h
+++ b/tools/imagetool.h
@@ -271,19 +271,22 @@ int rockchip_copy_image(int fd, struct image_tool_params *mparams);
  *  b) we need a API call to get the respective section symbols */
 #if defined(__MACH__)
 #include <mach-o/getsect.h>
+#include <mach-o/dyld.h>
 
-#define INIT_SECTION(name)  do {					\
+#define INIT_SECTION(name)	struct image_type_params		\
+	**__cat(__start_, name), **__cat(__stop_, name);		\
+	do {								\
 		unsigned long name ## _len;				\
-		char *__cat(pstart_, name) = getsectdata("__TEXT",	\
-			#name, &__cat(name, _len));			\
+		char *__cat(pstart_, name) = getsectdata("__DATA",	\
+			#name, &__cat(name, _len)) +			\
+			_dyld_get_image_vmaddr_slide(0);		\
 		char *__cat(pstop_, name) = __cat(pstart_, name) +	\
 			__cat(name, _len);				\
 		__cat(__start_, name) = (void *)__cat(pstart_, name);	\
 		__cat(__stop_, name) = (void *)__cat(pstop_, name);	\
 	} while (0)
-#define SECTION(name)   __attribute__((section("__TEXT, " #name)))
+#define SECTION(name)   __attribute__((section("__DATA, " #name)))
 
-struct image_type_params **__start_image_type, **__stop_image_type;
 #else
 #define INIT_SECTION(name) /* no-op for ELF */
 #define SECTION(name)   __attribute__((section(#name)))
```

cause define varliables(__start_image_type/__stop_image_type) in .h file is not an good idea here

svlobanov commented:
> changes is good, but I prefer do some small change:

it breaks compiling process

```
...
  HOSTCC  tools/imagetool.o
tools/imagetool.c:15:2: error: use of undeclared identifier '__start_image_type'; did you mean 'pstart_image_type'?
        INIT_SECTION(image_type);
                                         ^
```

ptpt52 commented:
> > changes is good, but I prefer do some small change:
> 
> it breaks compiling process
> 
> ```
> ...
>   HOSTCC  tools/imagetool.o
> tools/imagetool.c:15:2: error: use of undeclared identifier '__start_image_type'; did you mean 'pstart_image_type'?
>         INIT_SECTION(image_type);
>                                          ^
> ```

Please check, it build ok on my env.

svlobanov commented:
@ptpt52 you are right. I've sent a patch with your changes to upstream https://lists.denx.de/pipermail/u-boot/2021-December/468875.html

Let's wait for some time for their comments and decide what to do next

I'm moving this PR to draft status while waiting for a feedback from upstream members

svlobanov commented:
BTW, fakeroot upstream accepted macos arm64 changes (from tools/fakeroot/100-macos_compile_fix.patch and tools/fakeroot/500-macos_arm64_compile_fix.patch) so we will be able to remove both patches from OpenWRT after future release of fakeroot

https://salsa.debian.org/clint/fakeroot/-/merge_requests/13

aparcar commented:
The current patch works for me, thanks!

svlobanov commented:
> The current patch works for me, thanks!

upstream members don't reply to my message with the patch so I'm moving this PR to normal status to merge it to openwrt tree

@aparcar, can you post a message in u-boot mailing list that this patch works on your apple mac? 
thread: https://lists.denx.de/pipermail/u-boot/2021-December/468875.html
maillist: https://lists.denx.de/listinfo

aparcar commented:
Merged, thank you. I'll comment on the mailing list

kokroo commented:
@ptpt52 Still unable to compile x-wrt on macOS 11 with Apple Silicon. Have you ever tried it?

aparcar commented:
x-wrt != openwrt, please ask in the related issue tracker
